libssh-rs/src/lib.rs

```diff
@@ -274,6 +282,55 @@ impl Session {
         }
     }
 
+    unsafe extern "C" fn bridge_channel_open_request_auth_agent_callback(
+        session: sys::ssh_session,
+        userdata: *mut ::std::os::raw::c_void,
+    ) -> sys::ssh_channel {
+        let result = std::panic::catch_unwind(|| -> SshResult<sys::ssh_channel> {
+            let sess: &mut SessionHolder = &mut *(userdata as *mut SessionHolder);
+            assert!(
+                std::ptr::eq(session, sess.sess),
+                "invalid callback invocation: session mismatch"
+            );
+            let cb = sess
+                .channel_open_request_auth_agent_callback
+                .as_mut()
+                .unwrap();
+            let chan = unsafe { sys::ssh_channel_new(session) };
+            if chan.is_null() {
+                return Err(sess
+                    .last_error()
+                    .unwrap_or_else(|| Error::fatal("ssh_channel_new failed")));
+            }
+            match cb(Channel::new(&sess.outer.upgrade().unwrap(), chan)) {
+                // SAFETY: We steal a *mut sys::ssh_channel_struct here and let libssh
```

Comment:
I think the lifetime management around this is a bit sketchy; we cannot guarantee that this is safe as written; there's no way to guarantee that the callback will return the appropriate Error variant on failure.

As an alternative, how about we expose this as a two-callback API?
* First callback is `fn request_agent_callback(channel: &Channel) -> bool` to approve/deny the forward on a non-owning reference to the channel
* Second callback is triggered only when the first returns true and is something like `fn agent_forward_established(channel: Channel)` which passes ownership.

This way the careful internal grubbing around stuff can be contained to this callback and can't leak out to allow the Channel to be used unsafely.

For the destruction code path, I don't like the idea of partially destroying `Channel` and then allowing its Drop impl to run; it allows the object to exist in a partially destroy state that makes it conceptually harder to reason about.

I think an alternative strategy might look like this, which is a little bit more code, but it is reasonably well contained:

```rust
let channel = Channel::new(&sess.outer.upgrade()?, chan);
if request_agent_callback(&channel) {
   agent_forward_established(channel);
   return;
}

unsafe {
 // Disable callbacks during destruction
 // Maybe this could be moved to a pub(crate) method and shared
 // with the Drop impl?
 sys::ssh_remove_channel_callbacks(
     chan.chan_inner,
     chan._callbacks.as_mut());
 // release it
 sys::ssh_channel_free(chan.chan_inner); 
}
// Bypass the Drop impl
std::mem::forget(chan);
```

Reply:
> but I can't come up with a way to fix this without allowing the Channel to be used unsafely

Hm, maybe we can make a `struct UnlockedChannel` without `Send`, pass that to callbacks, and have a `upgrade()` that consumes it and returns back a normal `Channel`. I can do this refactor in another PR first if you prefer so.

Reply:
FWIW, I still don't think we can guarantee correct behavior; for example, if either callback panics then Channel::Drop will still run and deadlock.

An alternative approach might be to do something like this, using `Arc<Channel>` as a way to manage the lifetime. The callback would take `Arc<Channel>` rather than any other kind of reference to a channel. 

```rust
let mut channel_holder = None;

let result = std::panic::catch_unwind(|| {
   let channel = Arc::new(Channel::new(..)));
   channel_holder.replace(Arc::clone(&channel));
   cb(channel)
});

fn drop_if_only_ref(channel: Arc<Channel>) {
   if let Some(channel) = channel.into_inner() {
      drop_channel_unlocked(channel);
   }
}

match result {
  Err(err) => {
     drop_if_only_ref(channel_holder)
     std::ptr::null_mut()
  }
  Ok(Err(err)) => {
     drop_if_only_ref(channel_holder)
     std::ptr::null_mut()
  }
  Ok(Ok(chan)) => chan,
}
```

The big downside is that this would be the only API surface where Arc is a thing, which feels like an ergonomics wart, and I don't really like it.

Another option is again to split the API surface for this:

* callback to approve/reject the forward as I suggested earlier, accepting `&Channel`
* On approval, the channel is queued to a list of pending channels that is stored in `SessionHolder`.  This can be a `Vec<LockedChannel>` where `LockedChannel` is just the bits we need to remember to free up, which we move from the temporary `Channel` that we pass to the callback.  (Note that we may still need a variant of the above, to safely destruct it across a panic, but we can do so without an Arc):

```rust
struct LockedChannel {
  chan_inner: sys::ssh_channel,
  _callbacks: Box<sys::ssh_channel_callbacks_struct>,
  callback_state: Box<CallbackState>,
}
```

* Add `Session::pop_auth_agent_channel` method to retrieve and move a `LockedChannel` instance and convert it into a `Channel`.  In this context, there is no error path that can trigger and cause a deadlock on drop.

A consequence of this is that the application loop would need to periodically call this new method to detect and process these new channels, which isn't ideal, but is something that can be rolled into wezterm-ssh's SessionEvent reasonably well.

Reply:
I like the `Session::pop_auth_agent_channel` idea! I've just pushed a new version and it actually simplified both libssh-rs and wezterm-ssh side. It largely follows your idea, besides:

> callback to approve/reject the forward as I suggested earlier, accepting `&Channel`

We can't pass the callback a `&Channel`, it still can't do anything on it due to lock issues. Instead, the callback became rather dumb: `fn() -> bool`. API-wise, do we still need a callback here, or should we do `Session::enable_accept_agent_forward` and `Session::disable_accept_agent_forward`?

The only downside of not having a callback I can think of is, now we can't connect to ssh agent before accepting the forwarding request, so in case we can't reach to the agent, we accept the forwarding request and then immediately send EOF instead of refusing it. Should not really matter. WDYT?

Reply:
> API-wise, do we still need a callback here, or should we do `Session::enable_accept_agent_forward` and `Session::disable_accept_agent_forward`?

Ah. I'd assumed that there was something that the embedding application might want to do to check the channel, but there isn't really anything in the API for it call, so, yeah, I think this whole thing is a lot simpler if we just add a method to set a boolean and we can do away with the actual callback part of this.

So, the question is: do we default to enable or disable?  How does this relate to the corresponding ssh config option?
Does libssh even care about that?
If not, then I think the logical thing to do in the embedding application is to enable/disable based on the config.

Reply:
Done. This is now a trivial 50 lines change. Thanks for your guidance!

> Does libssh even care about that?

No. Yeah, just let embedding app control it.

Reply:
I like this a lot, thank you for working through this!